### PR TITLE
Enable Edgezone on CAPZ

### PR DESCRIPTION
**Pull request by mogliang**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -21,7 +21,7 @@ settings = {
     "kind_cluster_name": "capz",
     "capi_version": "v1.2.4",
     "cert_manager_version": "v1.1.0",
-    "kubernetes_version": "v1.23.9",
+    "kubernetes_version": "v1.25.0",
```
Comment (mogliang) [inline Tiltfile]: revert change

File: Tiltfile [hunk 2]
```diff
@@ -329,7 +329,9 @@ def deploy_worker_templates(template, substitutions):
     # programmatically define any remaining vars
     # "windows" can not be for cluster name because it sets the dns to trademarked name during reconciliation
     substitutions = {
-        "AZURE_LOCATION": "eastus",
+        "AZURE_LOCATION": "eastus2euap",
+        "AZURE_EXTENDEDLOCATION_TYPE": "EdgeZone",
+        "AZURE_EXTENDEDLOCATION_NAME": "microsoftrrdclab3",
```
Comment (mogliang) [inline Tiltfile:334]: is substitutions values can be overrided by system variable? what's the dev experience ?

File: api/v1beta1/types_class.go [hunk 3]
```diff
@@ -25,6 +25,9 @@ type AzureClusterClassSpec struct {
 
 	Location string `json:"location"`
 
+	// +optional
```
Comment (mogliang) [inline api/v1beta1/types_class.go:29]: comment

File: api/v1beta1/types_class.go [hunk 4]
```diff
@@ -52,6 +55,12 @@ type AzureClusterClassSpec struct {
 	CloudProviderConfigOverrides *CloudProviderConfigOverrides `json:"cloudProviderConfigOverrides,omitempty"`
 }
 
+type ExtendedLocationSpec struct {
```
Comment (mogliang) [inline api/v1beta1/types_class.go:59]: comment

File: azure/converters/extendedlocation.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+package converters
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+//  ExtendedLocationToNetworkSDK converts infrav1.ExtendedLocationSpec to network.ExtendedLocation.
```
Comment (mogliang) [inline azure/converters/extendedlocation.go]: remove unnecessary whitespace

File: azure/converters/extendedlocation.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+		return nil
+	} else {
+		return &network.ExtendedLocation{
+			Name: to.StringPtr(src.Name),
+			Type: network.ExtendedLocationTypes(src.Type),
+		}
+	}
+}
+
+//  ExtendedLocationToComputeSDK converts infrav1.ExtendedLocationSpec to compute.ExtendedLocation.
```
Comment (mogliang) [inline azure/converters/extendedlocation.go]: remove unnecessary whitespace

File: azure/scope/machine.go [hunk 7]
```diff
@@ -485,6 +488,9 @@ func (m *MachineScope) AvailabilitySet() (string, bool) {
 
 // AvailabilitySetID returns the availability set for this machine, or "" if there is no availability set.
 func (m *MachineScope) AvailabilitySetID() string {
+	if m.ExtendedLocation() != nil {
```
Comment (mogliang) [inline azure/scope/machine.go]: move condition check to m.AvailabilitySet()

File: azure/scope/managedcontrolplane.go [hunk 8]
```diff
@@ -151,6 +151,11 @@ func (s *ManagedControlPlaneScope) Location() string {
 	return s.ControlPlane.Spec.Location
 }
 
+// ExtendedLocation returns the control plane's extendedLocation.
+func (s *ManagedControlPlaneScope) ExtendedLocation() *infrav1.ExtendedLocationSpec {
+	return s.ControlPlane.Spec.ExtendedLocation
```
Comment (mogliang) [inline azure/scope/managedcontrolplane.go]: We don't implement EdgeZone for AzureManagedControlPlance in this PR.
Return Nil directly.

File: azure/services/publicips/client.go [hunk 9]
```diff
@@ -60,13 +61,14 @@ func (ac *AzureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (
 // It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
 // progress of the operation.
 func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "publicips.AzureClient.CreateOrUpdate")
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "publicips.AzureClient.CreateOrUpdate")
```
Comment (mogliang) [inline azure/services/publicips/client.go]: revert.
check current logging, print extendedlocation along with location

File: azure/services/publicips/client.go [hunk 10]
```diff
@@ -60,13 +61,14 @@ func (ac *AzureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (
...
 	defer done()
 
 	publicip, ok := parameters.(network.PublicIPAddress)
 	if !ok {
 		return nil, nil, errors.Errorf("%T is not a network.PublicIPAddress", parameters)
 	}
+	log.Info(fmt.Sprintf("qliang2: publicIP extendedlocation %v", publicip.ExtendedLocation))
```
Comment (mogliang) [inline azure/services/publicips/client.go]: revert

File: controllers/azurecluster_reconciler.go [hunk 11]
```diff
@@ -143,6 +144,14 @@ func (s *azureClusterService) getService(name string) (azure.ServiceReconciler,
 // setFailureDomainsForLocation sets the AzureCluster Status failure domains based on which Azure Availability Zones are available in the cluster location.
 // Note that this is not done in a webhook as it requires API calls to fetch the availability zones.
 func (s *azureClusterService) setFailureDomainsForLocation(ctx context.Context) error {
+	ctx, log, _ := tele.StartSpanWithLogger(ctx, "controllers.azureClusterService.Delete")
+
+	log.Info(fmt.Sprintf("qliang1: extendedlocation info: %v", s.scope.ExtendedLocation()))
```
Comment (mogliang) [inline controllers/azurecluster_reconciler.go]: remove logging

File: controllers/azuremachine_reconciler.go [hunk 12]
```diff
@@ -46,11 +46,16 @@ type azureMachineService struct {
 
 // newAzureMachineService populates all the services based on input scope.
 func newAzureMachineService(machineScope *scope.MachineScope) (*azureMachineService, error) {
+
```
Comment (mogliang) [inline controllers/azuremachine_reconciler.go]: revert

File: controllers/azuremachine_reconciler.go [hunk 13]
```diff
@@ -46,11 +46,16 @@ type azureMachineService struct {
...
 	cache, err := resourceskus.GetCache(machineScope, machineScope.Location())
 	if err != nil {
 		return nil, errors.Wrap(err, "failed creating a NewCache")
 	}
 
+	// If ExtendedLocation exists, remove Availability set service.
+	if machineScope.ExtendedLocation() != nil {
+		return newAzureMachineServiceNoAS(machineScope, cache)
```
Comment (mogliang) [inline controllers/azuremachine_reconciler.go]: inline

File: exp/api/v1beta1/azuremanagedcontrolplane_types.go [hunk 14]
```diff
@@ -61,6 +61,9 @@ type AzureManagedControlPlaneSpec struct {
 	// Location is a string matching one of the canonical Azure region names. Examples: "westus2", "eastus".
 	Location string `json:"location"`
 
+	// +optional
+	ExtendedLocation *infrav1.ExtendedLocationSpec `json:"extendedLocation,omitempty"`
```
Comment (mogliang) [inline exp/api/v1beta1/azuremanagedcontrolplane_types.go]: revert

File: templates/flavors/edgezone/patches/azure-extendedlocation.yaml [hunk 15]
```diff
@@ -0,0 +1,9 @@
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureCluster
+metadata:
+  name: ${CLUSTER_NAME}
+spec:
+  extendedLocation:
```
Comment (mogliang) [inline templates/flavors/edgezone/patches/azure-extendedlocation.yaml:7]: test name and type = empty string and check the behavor

File: templates/flavors/private/patches/azure-bastion.yaml [hunk 16]
```diff
@@ -6,3 +6,6 @@ metadata:
 spec:
   bastionSpec:
```
Comment (mogliang) [inline templates/flavors/private/patches/azure-bastion.yaml:7]: remove this file

File: templates/cluster-template-private.yaml [hunk 17]
```diff
@@ -27,6 +27,9 @@ metadata:
 spec:
   bastionSpec:
     azureBastion: {}
+  extendedLocation:
```
Comment (mogliang) [inline templates/cluster-template-private.yaml]: revert this file

File: templates/test/ci/cluster-template-prow-private.yaml [hunk 18]
```diff
@@ -39,6 +39,9 @@ spec:
         - ${AZURE_BASTION_SUBNET_CIDR}
         name: AzureBastionSubnet
         role: bastion
```
Comment (mogliang) [inline templates/test/ci/cluster-template-prow-private.yaml]: revert this file

File: templates/cluster-template-private-external.yaml [hunk 19]
```diff
@@ -0,0 +1,257 @@
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
```
Comment (mogliang) [inline templates/cluster-template-private-external.yaml]: exclude this file in pr

File: azure/converters/extendedlocation.go (hunk 5 above)
Comment (upxinxin) [reply azure/converters/extendedlocation.go]: Has removed

File: templates/flavors/edgezone/patches/azure-extendedlocation.yaml (hunk 15 above)
Comment (upxinxin) [reply templates/flavors/edgezone/patches/azure-extendedlocation.yaml:7]: When name or type is empty string, there will be "failed to create AzureCluster object" error and be blocked.

File: templates/flavors/private/patches/azure-bastion.yaml (hunk 16 above)
Comment (upxinxin) [reply templates/flavors/private/patches/azure-bastion.yaml:7]: Keep bastionSpec as origin private flavor has it.

File: templates/cluster-template-private.yaml (hunk 17 above)
Comment (upxinxin) [reply templates/cluster-template-private.yaml]: Has removed extendedLocation part.

File: api/v1beta1/types_class.go (hunk 4 above)
Comment (upxinxin) [reply api/v1beta1/types_class.go:59]: Has added comment

File: Tiltfile (hunk 2 above)
Comment (upxinxin) [reply Tiltfile:334]: In tiltfile, only sub, tenantID, clientID, clientsecret are read from tilt-settings.yaml, other parameters are read directly from tiltfile, not from system variable. So modify the parameters directly in tilefile.
